content/recipes/interoperability/creating-knowledge-graph-from-unstructured-text.md

```diff
@@ -0,0 +1,454 @@
+
+Here, we will outline an approach to store the results as an RDF graph by using the python [rdflib](https://github.com/RDFLib) library.
+
+rdflib allows the creation of entities with known URIs with the URIRef command.
+
+Also, one can create a custom namespace with new entities and relations. 
+
+```{note}
+For our audience interesting in labeled property graphs, the guidelines to store the results as a neo4j labeled property
+graph are available from [this tutorial](https://towardsdatascience.com/extract-knowledge-from-text-end-to-end-information-extraction-pipeline-with-spacy-and-neo4j-502b2b1e0754).
+```
+
+
+```python
+from rdflib import Graph
+from rdflib import URIRef, BNode, Literal, Namespace
+import json
+
+def Capitalise_underscore(relation):
+  """
+  Write the relations in Capitalized_And_Underscored_form
+  """
+  return relation.capitalize().replace(' ','_')
+
+def ncit_iri(item):
+  """
+  Get an iri of the first matching term from NCIT
+  """
+  try:
+    url = f"https://www.ebi.ac.uk/ols/api/search?q={item}&ontology=ncit"
+    data = pd.DataFrame(requests.get(url).json().get('response').get('docs'))
+    # Return the first id
+    return data["iri"][0]
+  except:
+    return 'id-less'
+
+#create the graph
+g=Graph()
+EX = Namespace('http://example.org./')
+
+#dataframe storing the trios for visual clarity
+relations = pd.DataFrame()
+
+for i, paper in enumerate(papers['PubmedArticle']):
+    print("{}) {}".format(i+1, paper['MedlineCitation']['Article']['ArticleTitle']))
+    #get abstract
+    abstract_text_json = json.dumps(papers['PubmedArticle'][i]['MedlineCitation']['Article']['Abstract']['AbstractText'])
+    abstract_text = ' '.join(json.loads(abstract_text_json))
+    
+    #take through coreference model
+    coref_text = coref(abstract_text)._.resolved_text
+    
+    #entity extraction and linking
+    doc = rel_ext(coref_text)
+    
+    #adding trios to the graph
+    for value, rel_dict in doc._.rel.items():
+      subject_iri = ncit_iri(rel_dict['head_span']['text'])
+      object_iri = ncit_iri(rel_dict['head_span']['text'])
+      if subject_iri != 'id-less':
+        subj = URIRef(subject_iri)
+      else:
+        subj = EX[rel_dict['head_span']['text']]
+      if object_iri != 'id-less':
+        obj = URIRef(object_iri)
+      else:
+        subj = EX[rel_dict['head_span']['text']]  
+      pred = EX[Capitalise_underscore(rel_dict['relation'])]
+      g.add((subj, pred, obj))        
+    df = pd.DataFrame.from_dict(doc._.rel).transpose()
+    df['subject_text'] = df.head_span.apply(lambda x: x['text'])
+    df['subject_id'] = df.head_span.apply(lambda x: x['id'])
+    df['object_text'] = df.tail_span.apply(lambda x: x['text'])
+    df['object_id'] = df.tail_span.apply(lambda x: x['id'])
+
+    df = df.drop(["head_span", "tail_span"], axis = 1)
+
+    relations = pd.concat([relations, df])
+```
+
+Finally, we can export the resulting graph in turtle (.ttl) format ...
```

Comment:
Why are there 3 dots after format? Did you want to add something there?

Reply:
Yes, here is the collab:
https://colab.research.google.com/drive/1La1oo3h9dAO5WGEgulZ6qTm_0gmhQoIn?authuser=1#scrollTo=Te0RwWgoziNV
I wonder if we should give a github link with this jupyter notebook or how to organize it better?

About three dots not sure, I guess Philippe wanted to add something.